tests/msc2403_test.go

```diff
@@ -0,0 +1,281 @@
+// +build msc2403
+
+package tests
+
+import (
+	"encoding/json"
+	"net/url"
+	"testing"
+
+	"github.com/matrix-org/complement/internal/b"
+	"github.com/matrix-org/complement/internal/client"
+	"github.com/matrix-org/complement/internal/must"
+	"github.com/tidwall/gjson"
+)
+
+// A reason to include in the request body when testing knock reason parameters
+var testKnockReason string = "Let me in... LET ME IN!!!"
+
+// TestKnocking tests sending knock membership events and transitioning from knock to other membership states.
+// Knocking is currently an experimental feature and not in the matrix spec.
+// This function tests knocking on local and remote room.
+func TestKnocking(t *testing.T) {
+	deployment := Deploy(t, "test_knocking", b.BlueprintFederationTwoLocalOneRemote)
+	defer deployment.Destroy(t)
+
+	// Create a client for one local user
+	aliceUserID := "@alice:hs1"
+	alice := deployment.Client(t, "hs1", aliceUserID)
+
+	// Create a client for another local user
+	bobUserID := "@bob:hs1"
+	bob := deployment.Client(t, "hs1", bobUserID)
+
+	// Create a client for a remote user
+	charlieUserID := "@charlie:hs2"
+	charlie := deployment.Client(t, "hs2", charlieUserID)
+
+	// Create a room for alice and bob to test knocking with
+	roomIDOne := alice.CreateRoom(t, struct {
+		Preset      string `json:"preset"`
+		RoomVersion string `json:"room_version"`
+	}{
+		"private_chat",      // Set to private in order to get an invite-only room
+		"xyz.amorgan.knock", // Room version required for knocking. TODO: Remove when knocking is in a stable room version
+	})
+
+	// Test knocking between two users on the same homeserver
+	knockingBetweenTwoUsersTest(t, roomIDOne, alice, bob, false)
+
+	// Create a room for alice and charlie to test knocking with
+	roomIDTwo := alice.CreateRoom(t, struct {
+		Preset      string `json:"preset"`
+		RoomVersion string `json:"room_version"`
+	}{
+		"private_chat",      // Set to private in order to get an invite-only room
+		"xyz.amorgan.knock", // Room version required for knocking. TODO: Remove when knocking is in a stable room version
+	})
+
+	// Test knocking between two users, each on a separate homeserver
+	knockingBetweenTwoUsersTest(t, roomIDTwo, alice, charlie, true)
+}
+
+func knockingBetweenTwoUsersTest(t *testing.T, roomID string, inRoomUser, knockingUser *client.CSAPI, federation bool) {
+	knockingUserID := knockingUser.UserID
+
+	t.Run("Knocking on a room with a join rule other than 'knock' should fail", func(t *testing.T) {
+		knockOnRoomWithStatus(t, knockingUser, roomID, "Can I knock anyways?", []string{"hs1"}, 403)
+	})
+
+	t.Run("Change the join rule of a room from 'invite' to 'xyz.amorgan.knock'", func(t *testing.T) {
+		inRoomUser.MustDo(
+			t,
+			"PUT",
+			[]string{"_matrix", "client", "r0", "rooms", roomID, "state", "m.room.join_rules", ""},
```

Comment:
Don't do this, use `CSAPI.SendEventSynced(t, roomID, b.Event)` to send state events into the room. This will wait for the response to come down `/sync` which is critical to reduce racey tests.